### ISSUE TITLE
How does parallelization in espnet work?

### ISSUE THREAD

**Issue by anuragkumar95** [espnet/espnet#3424]
I would like to understand how does espnet2 parallelizes the model and which mode (multithreaded vs multiprocess) will work for my usecase.

I would like to calculate the difference between the losses before and after training on a single minibatch. This works fine when I am running the model over a single GPU, however, I end up getting different values when I run espnet over multiple GPU's. 
```
1. loss_1 = model(batch) (in .eval() mode)
2. model(batch) (train mode)
3. loss_2 = model(batch) (in .eval() mode)
```
I have logged the losses, and I get different values over 2 GPUs. The losses have the above loss in the form (loss_1, loss_2).
```
{"losses": "(1057.0538330078125, 1056.30419921875)", "iepoch": 1, "iiter": 6}
{"losses": "(1082.687744140625, 1082.0916748046875)", "iepoch": 1, "iiter": 6}
```
Any suggestions why would it happen after doing .all_reduce()? If the two losses are returned I would expect them to be the same.

Thank you!

**Comment by sw005320** [comment]
It might be due to
1. specaugment or dropout
2. non-deterministic operation

For 1, please test it by disabling it.
For 2, we try to make the operation deterministic as much as possible but it still may exist.

**Comment by sw005320** [comment]
Oh, sorry `.eval() ` must avoid 1.

**Comment by anuragkumar95** [comment]
Is there a way to make sure this calculation happens only on one of the processes instead of all? Also, can you explain a bit more about (2)?

**Comment by sw005320** [comment]
For 2, there are some operations (e.g., atomic operations), which do not guarantee the same results. For example, it would even happen in the basic operation like conv2d (see the second note in https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html). 

I think we make `torch.backends.cudnn.deterministic = True`, but we might forget some parts.
I did not check it recently.

> Is there a way to make sure this calculation happens only on one of the processes instead of all? 
Frankly, I do not fully understand the issue.
Let me make time later, and I'll respond it.

**Comment by anuragkumar95** [comment]
To make the issue more clear, when the training is started in multiprocessing mode, the trainer spawns ngpu+1 processes. All the losses that are calculated have all_reduce() called. In my opinion, the sum of all the losses is copied to each process after this. Since there is a randomness due to the reasons stated above, an average of loss is calculated and passed for backpropagation. What I need to calculate are the losses before and after the training step on a minibatch. To do this without affecting the gradients, I use the eval() mode to calculate the loss. Does the loss calculation in the eval mode happens in a similar way?  Aren't dropouts disabled when your model is in eval() mode?  

My first guess was that since there are multiple processes with different examples, the losses across the process will be different, however, they should be the same after all_reduce() and average.  Since this isn't the case, is there anything I am missing out on?? Is there a way to control this operation to be done on only one of the processes and not the others?

**Comment by sw005320** [comment]
Thanks for the clarification, but frankly I don't know what happens...
Yes, the loss calculation in the eval mode happens in a similar way and it would not apply dropout and specaug...
